adaboost/cuda/core/function_params.cu

```diff
@@ -0,0 +1,21 @@
+#ifndef ADABOOST_UTILS_FUNCTION_PARAMS_CU
+#define ADABOOST_UTILS_FUNCTION_PARAMS_CU
+
+namespace adaboost
+{
+    namespace cuda
+    {
+        namespace core
+        {
+            __device__ float cube(float x)
+            {
+                return x*x*x;
+            }
+            
+            __device__  adaboost::cuda::core::func_t<float,float> p_func_here = adaboost::cuda::core::cube;
```

Comment:
We will be required to declare different pointers for different types, right? This seems pretty challenging problem to solve.

Reply:
For different data types we can use a templates. I have just pushed a commit in which I made the changes, please see that. The tests are passing for that.

The only problem is if we want multiple functions to be called, we will need to make separate variables of type `func_t` for each. I tried to make an array of type `func_t`, but here also it is giving me an `invalid device symbol error`. So looks like that is not possible. Possible workarounds are:

1. Make a separate `func_t` data type variable for each function. Then we can use switch case to choose between them. Switch case will be implemented in `Argmax` from `operations_impl.hpp` file.
2. The second option is to instead pass another argument to the current `cube` function. Here we can use switch case again. So in this method the checking will be done every time the function is called, but it will eliminate the need for calling separate pointers. 

 Please let me know your thoughts on this.

Reply:
The problem is we want to define custom functions which should be executed on every element of a vector on a GPU. The custom functions are to be executed on GPU so, we cannot use pointers to these functions in host code. But one point to be noted here is that, `Argmax` will not be exposed in the public API of AdaBoost algorithm, so making `Argmax` a `__global__` function will not hurt but we cannot make it a `__device__` function and hence, function pointers will keep creating problems. 
I suggest the following solution,

1. Remove usage of `__device__` function pointers in parameter list of functions. Instead, some number or some unique id associated with a function can be passed to `Argmax` function and that should be used by CUDA kernel to execute the required function on GPU.
2. Call `__device__` functions from the CUDA kernel. The `__device__` functions are by default inlined, so it should work.
3. Define different custom functions to be applied on all the elements of the vector inside the kernel should be defined in the same file as of `Argmax` and `Sum` or whatever GPU operations. The function id received through Argmax can be used in some if-else ladder and correct function can be called.

Reply:
Please let me know if something is not clear.

Reply:
Please see the recent commit. I have tried to incorporate the required changes. Some important points I want to mention are:
1. I tried to pass the `__device__` function directly to the kernel, but it is not possible. If we want to avoid `cudaMemcpyFromSymbol` the only way we can do that is by including the if else ladder inside the `__global__` function every time we want to call `func_ptr`. This code will not only be cumbersome, but also slower since each thread will have to go through the entire ladder. Hence I have retained the use of `cudaMemcpyFromSymbol`.
2. `function_params` file is where all the `__device__` functions and static pointers to them are defined. I thought it would become too confusing to put them in `operations_impl.cu`. Please let me know if you want to include them in `operations_impl.cu`.
3. Current `Argmax` function has two template data types: `data_type_vec` and `data_type_ret`. I tried to remove the first argument in the Argmax function but it threw an error. Reason being it could not resolve what `data_type_ret` was since there remained no parameter in the function that could decide what it was. The only workaround I could figure out for this was to declare a dummy argument `val` of type `data_type_ret *`, and then freeing the variable after the call. This is the only way for the compiler to know what `data_type_ret` is. And it is necessary to have this template class since later on it is passed to the static pointer to device function (see line 353 in `operations_impl.cu`) and also to the `__global__` function. If you know of any other solutions to this please do tell.